### ISSUE TITLE
Magento 2.1.3 Delete Products error

### ISSUE THREAD

**Issue by adityag186** [magento/magento2#8159]
### Preconditions

1. Magento 2.1.3
2. 

### Steps to reproduce

1. Go to System->Import
2. Entity Type ->Prpducts
3. behaviour -> Delete
4. File To import- Please find attached file
5. Check Data
6. Import Data

### Expected result

1. Products should be deleted

### Actual result

1. Below error received

General system exception happened
Additional data: SQLSTATE[42000]: Syntax error or access violation: 1142 TRIGGER command denied to user 'atlan621_magen51'@'localhost' for table 'mgcv_catalog_product_entity', query was: DELETE FROM `mgcv_catalog_product_entity` WHERE (entity_id IN ('37286', '37287', '37288', '37289', '37290', '37291', '37292', '37293', '37294', '37295', '37296', '37297', '37298', '37299', '37300', '37301', '37302', '37303', '37304', '37305', '37306', '37307', '37308', '37309', '37310', '37311', '37312', '37313', '37314', '37315', '37316', '37317', '37318', '37319', '37320', '37321', '37322', '37323', '37324', '37325', '37326', '37327', '37328', '37329', '37330', '37331', '37332', '37333', '37334', '37335'))

**Comment by nikoelgatito** [comment]
@adityag186 Please make sure that user `atlan621_magen51` has all privileges on your Magento database.

This error message basically says that the database user `atlan621_magen51` is not allowed to use the `TRIGGER` command on this database.

**Comment by adityag186** [comment]
Hi,

Can you suggest some steps that I need to do for assigning all priviledges.

**Comment by nikoelgatito** [comment]
@adityag186 If you are using CPanel you may navigate to `MySQL Databases` inside the `DATABASES` section or search for it. Then look for the `Current Databases` section and on the same row as your Magento database under the column `Privileged Users` click on the user `atlan621_magen51`.

On the next page titled `Manage User Privileges` , simply check the top left checkbox for `ALL PRIVILEGES` and finally click on the `Make Changes` button at bottom left of the page.

**You may also refer to this CPanel documentation page:**
[MySQL Databases](https://documentation.cpanel.net/display/ALD/MySQL+Databases)

**If you are not using CPanel you may use a tutorial like the following:**
[HOW DO I GRANT PRIVILEGES IN MYSQL?](https://mediatemple.net/community/products/dv/204404494/how-do-i-grant-privileges-in-mysql)

**Comment by adityag186** [comment]
Hi,

My database name is devatlan_newdb and user name is also devatlan_newdb. However, magento shows error with username atlan621_magen51. This is the user for my live site. devatlan_newdb is the user for my development site ie. the site I am currently facing an error. Why is my site trying to use atlan621_magen51 instead of devatlan_newdb. I have made the required changes in the env.php file of database, database user in app/etc.

Are there any other things I need to do?

**Comment by nikoelgatito** [comment]
@adityag186 To be honest when I firstly encounter your bug report I was looking for a similar behavior as we also have an issue where a dev site is trying to use username from the production site database which are both hosted under the same server space. I did not mention it at the begining as in our case the error is not reproduced the same way and you had not mentionned anything regarding the env.php file.

In my case both `env.php` (dev and prod installations) are well configured and I even tried to remove renamed the files without effect. This looks like some caching/symlinking issue, I have refreshed all Magento cache even generation folders without luck. I thought about a problem with MySQL but I'm not experiencing any similar issues anywhere else. To reproduce the issue on our side, it is triggered while trying to save some categories and the missing priviledge is also `TRIGGER`. Our dev database is almost exactly the same as the production one. Usually Magento uses the same resource classes to initiate connections to the database, I'm not sure to understand how a different user could be used for a subsequent connection. Now I'm thinking, as we basically imported the very same database for both site only under different names, changing the `base_url` `core_config` settings, this problem might be related with the database definer. I'll definately have to dig deeper into this issue as this may cause disastrous issues.

Did you by any chance import both databases using the same source file?

**Comment by adityag186** [comment]
Hi,

You described the problem correctly. Even checkout process is giving error on the test website even though it is the same exact copy of live website.

I did not really understand the below question you asked.
Did you by any chance import the both databases using the same source file?

If i am interpreting the question correctly, then my response would be -: For the database, I think we just took a backup of database of live site and put it with the test site.

**Comment by nikoelgatito** [comment]
@adityag186 That was exactly the meaning of my question, we did the very same which is a standard procedure.

I've just tried to temporarily delete and add both database users back without success.

**There is the MySQL error output:**
```
[Zend_Db_Statement_Exception]
  SQLSTATE[42000]: Syntax error or access violation: 1142 TRIGGER command den
  ied to user 'prod_user'@'localhost' for table 'catalog_cate
  gory_entity', query was: UPDATE `catalog_category_entity` SET `attribut
  e_set_id` = ?, `parent_id` = ?, `path` = ?, `position` = ?, `level` = ?, `c
  hildren_count` = ? WHERE (entity_id = '1')
```

Unfortunately I think that the best solution will be to put the production site in maintenance mode to prevent traffic from updating the database, then do a new mysql dump. This time making sure to replace ALL `DEFINER`. A `DEFINER` basically define the account (mysql user) who has all priviledges to a database. 

**When doing a mysqldump, the output file will have contains some `DEFINER` which look like this:**
/*!50017 DEFINER=`user`@`localhost` */

The goal is to replace these by the right user which will be the one having all priviledges over your MySQL database once restored. What is probably causing our issues is that a definer might have been missed. You will also want to pay attention not to replacing other similar string by mistake while searching/replacing through your MySQL dump file.

**[ALWAYS BACKUP FIRST]**
Once this is done, I would remove the dev database and restore the dump file to the dev database again.

I will give it a try as soon as I find some time and post the result here.

**This StackExchange question should also  be helpful:**
[SQL Syntax error or access violation: 1142 TRIGGER command denied to user](/questions/90003/sql-syntax-error-or-access-violation-1142-trigger-command-denied-to-user)

**Comment by adityag186** [comment]
Hi,

Thanks for the information. Please do update me when you are done with the process.

**Comment by nikoelgatito** [comment]
@adityag186 **- UPDATE - It did work perfectly.**

## There are the steps to resolve this issue:

**0. Always do a safe backup of your database first,**

**1. Simply do a mysqldump to a file;**
```
mysqldump -u[username] -p [db_name] > [db_name].sql
```

**2. Search and replace any `DEFINER` occurrence in the dump file with the database user/owner;**
A `DEFINER` should look like this: ```/*!50017 DEFINER=`user`@`localhost`*/```

Simply replace `user` by your user/owner. (You may also simply remove all `DEFINER` entirely, on import all `TRIGGER` `DEFINER` should default to the user/owner.

_Tip: In case of a very large database, the dump file might end up too large to be open with a usual text editor, VIM can be helpful in these circumstances._

**3. Drop the database and recreate it with the right user (same as `DEFINER`, use same password and should have all privileges);**
```
DROP DATABASE [db_name];
CREATE DATABASE [db_name];
CREATE USER '[username]'@'localhost' IDENTIFIED BY '[password]';
GRANT ALL PRIVILEGES ON [db_name].* TO '[username]'@'localhost';
FLUSH PRIVILEGES;
```

**4. Import the dump file to the new database;**
```
mysql -u[username] -p [db_name] < [db_name].sql
```
**5. You may remove the dump file;**
```
rm [db_name].sql
```

### Done, the issue should be resolved!

**Comment by Stas94** [comment]
@adityag186 We cannot reproduce this issue as described. Please provide the detailed steps we must follow to reproduce this issue. In addition, identify the web server you are running, the versions of PHP and MySQL, and any other information needed to reproduce your issue.

**Comment by veloraven** [comment]
According to [contributor guide](http://devdocs.magento.com/guides/v2.1/contributor-guide/contributing.html), tickets without response for two weeks should be closed.
If this issue still reproducible please feel free to create the new one: format new issue according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines): with steps to reproduce, actual result and expected result and specify Magento version.
